### DISCUSSION TITLE
attempt to index a nil value

### DISCUSSION THREAD

**Discussion by DRBAUS** [RomanKubiak/ctrlr#467] [Q&A]
I am running into an attempt to index a nil value on two scripts: the loadState and Highlight Matrix (the wavesample select dropdown menu) and I don't understand why. 

In an earlier version I had the 'store parameter values for each wavesample' working but only for values that were physically changed (by mouse or numeric input) but not when the value was changed remotely by another modulator so I decided to change that but now 'attempt to index a nil value' even though the sampleValueData table is identical. 

Any suggestions where/what to look at?
[ENSONIQ EPS-16 PLUS_version_129.zip](https://github.com/RomanKubiak/ctrlr/files/8748437/ENSONIQ.EPS-16.PLUS_version_129.zip)

**Comment by dnaldoog** [comment]
Hi @DRBAUS 

You need a blocking script at the head of those functions to stop them loading.
`if block then return end`

In the `loadState`  line 21 there is a missing modulator name that it is looking for. You'll have to debug that one.

[ENSONIQ EPS-16 PLUS_1_72_block.zip](https://github.com/RomanKubiak/ctrlr/files/8748462/ENSONIQ.EPS-16.PLUS_1_72_block.zip)

**Reply by DRBAUS** [reply]
Oh...thanks. This solves both.

**Reply by DRBAUS** [reply]
Okay...the same issue keeps coming up.

I adapted some scripts and had it working flawlessly earlier. For some reason Ctrlr crashed when I pressed the BOOST button. I restarted, reloaded the panel and the 'attempt to index a nil value' keeps coming up on all scripts (4) that use this same line: sampleValueData["w"..set][name]=value

[ENSONIQ EPS-16 PLUS_135 nil value.zip](https://github.com/RomanKubiak/ctrlr/files/8790013/ENSONIQ.EPS-16.PLUS_135.nil.value.zip)

**Reply by dnaldoog** [reply]
> Okay...the same issue keeps coming up.
> 
> I adapted some scripts and had it working flawlessly earlier. For some reason Ctrlr crashed when I pressed the BOOST button. I restarted, reloaded the panel and the 'attempt to index a nil value' keeps coming up on all scripts (4) that use this same line: sampleValueData["w"..set][name]=value
> 

This should fix the problem.

You were trying to access the global variable 0 which was set to zero. When you changed the waveform you were always trying to access `sampleValueData[w0][name]` and such a table element doesn't exist - only w1-w8. 

You would notice that if you first selected a waveform from the wave select combo before doing anything else, your panel was working - because the global variable was set properly. Solutions to that would be to always ensure the global variable sits between 1-8 with something like `local set=jlimit(1,8,panel:getGlobalVariable(0))` and you could do that:

[ENSONIQ EPS-16 PLUS_1_72_using jlimit_2022-05-28_21-29.zip](https://github.com/RomanKubiak/ctrlr/files/8791133/ENSONIQ.EPS-16.PLUS_1_72_using.jlimit_2022-05-28_21-29.zip)


or I offer another solution: Either are good - you choose: 🤔

In this version, I changed the waveform select button to write `modulatorCustomIndex` variable 1-8 for each slider,combo etc residing in the `sampleValueData` table.

So instead of:

<pre>
local name=L(mod:getName())
local set=panel:getGlobalVariable(0) or 1
sampleValueData["w"..set][name]=value
sName = L(mod:getName())
</pre>
it is now 
<pre>
if source ~= 4 then return end
local name=L(mod:getName())
local set=mod:getProperty("modulatorCustomIndex")
sampleValueData["w"..set][name]=value
sName=name 
</pre>

With `sName=name ` I haven't changed them all, but either will work.


The combo code is changed from

<pre>
    panel:setGlobalVariable(0, value + 1)
    local key = "w" .. value + 1
    for k, v in pairs(sampleValueData[key]) do
        panel:getModulatorByName(k):setModulatorValue(tonumber(v), false,true,false) 
    end --  loop
</pre>

becomes

<pre>
    local key = "w" .. value + 1
    for k, v in pairs(sampleValueData[key]) do
        panel:getModulatorByName(k):setValue(tonumber(v), false) 
		panel:getModulatorByName(k):setProperty("modulatorCustomIndex",tostring(value + 1),false)
    end --  loop
</pre>


[ENSONIQ EPS-16 PLUS_1_72_modulatorCustomIndexVersion_2022-05-28_20-25.zip](https://github.com/RomanKubiak/ctrlr/files/8791046/ENSONIQ.EPS-16.PLUS_1_72_modulatorCustomIndexVersion_2022-05-28_20-25.zip)

----
Also, just a note for future - it's probably better to name your functions using `highLightMatrix= function()` not `HIGHLIGHTMATRIX = function()`  as a matter of accepted style, or I often do `high_light_matrix= function()`  because for me it is sometimes more readable. I think one shouldn't mix and match, but I do nevertheless! 😊

**Reply by dnaldoog** [reply]
Also my apologies for one faulty piece of code

`local set=panel:getGlobalVariable(0) or 1`

Because `panel:getGlobalVariable(0)` always exists, `or 1` is never reached, so in effect does nothing. Again to check that `panel:getGlobalVariable(0)` is always 1-8 we need another check:

<pre>
local set=jlimit(1,8,panel:getGlobalVariable(0))
</pre>

or

<pre>
local set=panel:getGlobalVariable(0)
if set  ==  0 then
set=1
end
</pre>

… but the latter is a bit clunky, especially if  `set=panel:getGlobalVariable(0)` somehow set itself to 9. It would throw an error
